### PR TITLE
add support for metadata in tracking

### PR DESCRIPTION
**Pull request by tuanddd**

### CODE REVIEW

**Comment by tuanddd** [comment]
Not sure if the change in `PosNegTracking` is necessary so just put your comment and will remove it.

File: packages/hooks/src/ContextProvider/controllers/tracking/ClickTracking.ts [hunk 1]
```diff
@@ -9,13 +9,14 @@ export class ClickTracking extends Tracking {
    * @param field Field to use for click token generation.
    * @param qParam Value to use for full-text query param.
```
Comment (jonathaningram) [inline packages/hooks/src/ContextProvider/controllers/tracking/ClickTracking.ts:10]: This is out of date.

File: packages/hooks/src/ContextProvider/controllers/tracking/ClickTracking.ts [hunk 2]
```diff
@@ -9,13 +9,14 @@ export class ClickTracking extends Tracking {
...
    */
-  constructor(field = 'url', qParam = 'q') {
+  constructor(field = 'url', metadata = { qParam: 'q' }) {
```
Comment (jonathaningram) [inline packages/hooks/src/ContextProvider/controllers/tracking/ClickTracking.ts:12]: Do we need `qParam` to go in `metadata` if we're just going to deference it to take it back out?

File: packages/hooks/src/ContextProvider/controllers/tracking/PosNegTracking.ts [hunk 3]
```diff
@@ -8,10 +8,10 @@ export class PosNegTracking extends Tracking {
    * Construct a PosNegTracking instance.
    * @param field Field to use for click token generation.
```
Comment (jonathaningram) [inline packages/hooks/src/ContextProvider/controllers/tracking/PosNegTracking.ts:9]: This is out of date.

File: .changeset/nervous-schools-cry.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+'@sajari/react-hooks': major
+---
+
+Allow for passing in metadata when using tracking constructors
```
Comment (jonathaningram) [inline .changeset/nervous-schools-cry.md:5]: Can we include a before and after snippet showing how to use the class before and after?

**Review by jonathaningram — COMMENTED** [review]
It looks like we don't have any tests for this since there are no modified tests in the PR. Any ideas how we could add a test for some of this tracking code?

**Comment by tuanddd** [comment]
Just missing the tests, which I'll figure out how to write them.

File: packages/hooks/src/ContextProvider/controllers/tracking/ClickTracking.ts [hunk 5]
```diff
@@ -7,15 +7,15 @@ export class ClickTracking extends Tracking {
   /**
    * Construct a ClickTracking instance.
    * @param field Field to use for click token generation.
-   * @param qParam Value to use for full-text query param.
+   * @param metadata Metadata fields.
    */
-  constructor(field = 'url', qParam = 'q') {
+  constructor(field = 'url', metadata: Record<string, string> = { qParam: 'q' }) {
```
Comment (jonathaningram) [inline packages/hooks/src/ContextProvider/controllers/tracking/ClickTracking.ts:12]: Is this metadata?

File: packages/hooks/src/ContextProvider/controllers/tracking/ClickTracking.test.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { TrackingType } from '@sajari/sdk-js';
+
+import { EVENT_TRACKING_RESET } from '../../events';
+import { ClickTracking } from './ClickTracking';
+
+describe('ClickTracking', () => {
+  it('should work normally', () => {
+    const clickTracking = new ClickTracking('id', 'q', { qParam: 'query', cartId: '12345' });
```
Comment (jonathaningram) [inline packages/hooks/src/ContextProvider/controllers/tracking/ClickTracking.test.ts:8]: delete `qParam: 'query', `

File: packages/hooks/src/ContextProvider/controllers/tracking/PosNegTracking.test.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import { TrackingType } from '@sajari/sdk-js';
+
+import { EVENT_TRACKING_RESET } from '../../events';
+import { PosNegTracking } from './PosNegTracking';
+
+describe('PosNegTracking', () => {
+  it('should work normally', () => {
+    const posNegTracking = new PosNegTracking('id', { qParam: 'query', cartId: '12345' });
```
Comment (jonathaningram) [inline packages/hooks/src/ContextProvider/controllers/tracking/PosNegTracking.test.ts:8]: delete `qParam: 'query', `

**Review by jonathaningram — COMMENTED** [review]
Pretty sure this is no longer a breaking change, agree? We're adding another constructor arg so is a minor bump but not a major bump.

File: .changeset/nervous-schools-cry.md [hunk 8]
```diff
@@ -0,0 +1,21 @@
+---
+'@sajari/react-hooks': minor
+---
+
+Allow for passing in object as second parameter when using tracking constructors
```
Comment (jonathaningram) [inline .changeset/nervous-schools-cry.md:5]: Second is not quite right. Maybe:

Allow callers to override the metadata when creating tracking objects

File: .changeset/nervous-schools-cry.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+```js
+new PosNegTracking('url');
+
+new ClickTracking('url', 'q');
+```
+
+After:
+
+```js
+new PosNegTracking('url', { qParam: 'q', foo: 'bar', ... })
```
Comment (jonathaningram) [inline .changeset/nervous-schools-cry.md:18]: qParam shouldn't be in the metadata

**Review by jonathaningram — COMMENTED** [review]
I think aside from the comments, should be good.

**Review by jonathaningram — APPROVED** [review]

**Pull request merged**